services/121-service/test/payment/do-payment-fsp-onafriq.test.ts

```diff
@@ -0,0 +1,242 @@
+  });
+
+  it('should give error via callback based on magic phonenumber', async () => {
+    // Arrange
+    registrationOnafriq.phoneNumber = '24300000002'; // this magic number is configured in mock to return an error on callback
+    await importRegistrations(programId, [registrationOnafriq], accessToken);
+    await awaitChangeRegistrationStatus({
+      programId,
+      referenceIds: [registrationOnafriq.referenceId],
+      status: RegistrationStatusEnum.included,
+      accessToken,
+    });
+    const paymentReferenceIds = [registrationOnafriq.referenceId];
+
+    // Act
+    const doPaymentResponse = await doPayment({
+      programId,
+      paymentNr: payment,
+      amount,
+      referenceIds: paymentReferenceIds,
+      accessToken,
+    });
+
+    // wait for non-waiting transactions only, to make sure callback came in
+    await waitForPaymentTransactionsToComplete({
+      programId,
+      paymentReferenceIds,
+      accessToken,
+      maxWaitTimeMs: 4_000,
+      completeStatusses: [
+        TransactionStatusEnum.success,
+        TransactionStatusEnum.error,
+      ],
+    });
+
+    // Assert
+    const getTransactionsBody = await getTransactions({
+      programId,
+      paymentNr: payment,
+      registrationReferenceId: registrationOnafriq.referenceId,
+      accessToken,
+    });
+
+    expect(doPaymentResponse.status).toBe(HttpStatus.ACCEPTED);
+    expect(doPaymentResponse.body.applicableCount).toBe(
+      paymentReferenceIds.length,
+    );
+    expect(getTransactionsBody.body[0].status).toBe(
+      TransactionStatusEnum.error,
+    );
+    expect(getTransactionsBody.body[0].errorMessage).toMatchSnapshot();
+  });
+
+  it('should not update transaction on a `duplicate thirdPartyTransId error` API response', async () => {
+    // Arrange
+    // NOTE 1: we use a magic phone number here that is configured in the mock to return a duplicate thirdPartyTransId error on request.
+    // We use this as we cannot actually easily test a duplicate thirdPartyTransId error in the mock.
+    registrationOnafriq.phoneNumber = '24300000001';
+    await importRegistrations(programId, [registrationOnafriq], accessToken);
+    await awaitChangeRegistrationStatus({
+      programId,
+      referenceIds: [registrationOnafriq.referenceId],
+      status: RegistrationStatusEnum.included,
+      accessToken,
+    });
+    const paymentReferenceIds = [registrationOnafriq.referenceId];
+
+    // Act
+    const doPaymentResponse = await doPayment({
+      programId,
+      paymentNr: payment,
+      amount,
+      referenceIds: paymentReferenceIds,
+      accessToken,
+    });
+
+    // Assert
+    // NOTE 2: we cannot wait for transactions to have a certain status, because we want to test that the transaction stays on 'waiting'.
+    // We also assert that no callback comes in, so we must give some time for the callback to potentially come in, as the assertion is not valuable otherwise.
+    await waitFor(1_000);
```

Comment:
Should we do:
```suggestion

    // Assert
    // NOTE 2: we cannot wait for transactions to have a certain status, because we want to test that the transaction stays on 'waiting'.
    // We also assert that no callback comes in, so we must give some time for the callback to potentially come in, as the assertion is not valuable otherwise.
    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      accessToken,
      maxWaitTimeMs: 4_000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
        TransactionStatusEnum.waiting,
      ],
    });
    await waitFor(1_000);
```